src/ts/forwarders/proxyAccountFowarder.ts

```diff
@@ -74,227 +76,167 @@ export class ProxyAccountForwarder extends Forwarder<
-   * @param data The target contract, value (wei) to send, and the calldata to execute in the target contract
-   */
-  protected getEncodedCallData(data: RequiredTo<ProxyAccountCallData>) {
-    return defaultAbiCoder.encode(
-      ["uint", "address", "uint", "bytes"],
-      [
-        data.callType ? data.callType : CallType.CALL,
-        data.to,
-        data.value ? data.value : 0,
-        data.data ? data.data : "0x",
-      ]
-    );
   }
 
-  /**
-   * Returns the encoded calldata for creating a proxy contract
-   * No need for ForwardParams as no signature is required in ProxyAccountDeployer
-   * @returns The proxy deployer address and the calldata for creating proxy account
-   * @throws If the proxy account already exists
-   */
-  public async createProxyContract(): Promise<MinimalTx> {
-    const callData = this.proxyDeployer.interface.functions.createProxyAccount.encode(
-      [await this.signer.getAddress()]
-    );
+  private defaultRevertableCallData(
+    data: RevertableProxyAccountCallData
+  ): Required<RevertableProxyAccountCallData> {
+    return {
+      ...this.defaultCallData(data),
+      revertOnFail: data.revertOnFail ? data.revertOnFail : false,
+    };
+  }
 
-    // 115k gas inc the transaction cost.
+  protected toCallData(
+    initCode: string,
+    extraData: string,
+    value: BigNumberish
+  ): ProxyAccountCallData {
     return {
-      to: this.proxyDeployer.address,
-      data: callData,
+      ...this.encodeForDeploy(initCode, extraData, value),
+      value: "0",
+      callType: CallType.DELEGATE,
     };
   }
 
-  /**
-   * Encodes the forward parameters such that it can be included in
-   * an Ethereum Transaction's data field.
-   * @param params Forward Parameters
-   */
-  protected async encodeSignedMetaTransaction(
-    params: ForwardParams
-  ): Promise<MinimalTx> {
-    const proxyAccount = new ProxyAccountFactory(this.signer).attach(params.to);
+  protected toBatchCallData(
+    initCode: string,
+    extraData: string,
+    value: BigNumberish,
+    revertOnFail?: boolean
+  ): RevertableProxyAccountCallData {
+    return {
+      ...this.encodeForDeploy(initCode, extraData, value),
+      value: "0",
+      callType: CallType.DELEGATE,
+      revertOnFail: revertOnFail || false,
+    };
+  }
 
-    const data = proxyAccount.interface.functions.forward.encode([
-      {
-        to: params.target,
-        value: params.value,
-        data: params.data,
-        callType: params.callType,
+  public decodeTx(data: string) {
+    const proxyAccount = new ProxyAccountFactory(this.signer).attach(
+      this.address
+    );
+    const parsedTransaction = proxyAccount.interface.parseTransaction({
+      data,
+    });
+    const functionArgs: {
+      _metaTx: Required<ProxyAccountCallData>;
+      _replayProtection: string;
+      _replayProtectionAuthority: string;
+      _signature: string;
+    } = {
+      _metaTx: {
+        to: parsedTransaction.args[0][0],
```

Comment:
So parseTransaction lets you separate it back to the arguments of the call?

Maybe we should have a better name than function args? You could easily return a ProxyCallData and a MinimalTx here?

Reply:
Maybe not a minimaltx? Not sure what data structure it should be?

Reply:
I was thinking that the reason you would use this function is because you would want to supply the data to the contract yourself - so I've named them all the same as they are in the contract.

Reply:
ok i thought it wouild be like const { functionArgs } = thisFunction(), but its just returning it and you name it whatever you want externally